File: src/main/connection/protocol-parser.ts
/**
 * 协议层状态机
 * 仅处理协议级解析（Telnet IAC、自定义 OSC），不处理 VT 渲染序列（由 xterm.js 处理）
 */

import { decodeBufferForTerminal } from './c1-convert'

/** 协议解析器状态 */
export const ParserState = {
  DATA: 0,        // 正常数据透传
  IAC: 1,         // 遇到 0xFF，等待命令字节
  IAC_DO: 2,      // IAC DO，等待选项字节
  IAC_DONT: 3,    // IAC DONT
  IAC_WILL: 4,    // IAC WILL
  IAC_WONT: 5,    // IAC WONT
  IAC_SB: 6,      // IAC SB，子协商开始
  IAC_SB_DATA: 7, // 子协商数据，收集到 IAC SE
  IAC_SB_IAC: 8,  // 子协商中遇到 IAC，等待 SE 或 IAC
} as const
export type ParserState = typeof ParserState[keyof typeof ParserState]

/** Telnet IAC 常量 */
const IAC = 255
const DO = 253
const DONT = 254
const WILL = 251
const WONT = 252
const SB = 250
const SE = 240

/** Telnet 选项常量 */
const OPT_ECHO = 1
const OPT_SUPPRESS_GO_AHEAD = 3
const OPT_NAWS = 31

/** 协议事件 */
export interface ProtocolEvents {
  data: string
  iacDo: number
  iacDont: number
  iacWill: number
  iacWont: number
  iacSb: { option: number; data: Buffer }
}

type EventCallback<T> = (value: T) => void

export class ProtocolParser {
  private state = ParserState.DATA
  private sbOption = 0
  private sbData: number[] = []
  private cleanBytes: number[] = []
  private callbacks: { [K in keyof ProtocolEvents]?: EventCallback<ProtocolEvents[K]>[] } = {}

  /** 输入原始字节流，返回剥离 IAC 后的干净 UTF-8 数据 */
  feed(raw: Buffer): string {
    for (let i = 0; i < raw.length; i++) {
      const byte = raw[i]

      switch (this.state) {
        case ParserState.DATA:
          if (byte === IAC) {
            this.state = ParserState.IAC
          } else {
            this.cleanBytes.push(byte)
          }
          break

        case ParserState.IAC:
          if (byte === IAC) {
            this.cleanBytes.push(IAC)
            this.state = ParserState.DATA
          } else if (byte === DO) {
            this.state = ParserState.IAC_DO
          } else if (byte === DONT) {
            this.state = ParserState.IAC_DONT
          } else if (byte === WILL) {
            this.state = ParserState.IAC_WILL
          } else if (byte === WONT) {
            this.state = ParserState.IAC_WONT
          } else if (byte === SB) {
            this.state = ParserState.IAC_SB
          } else {
            this.state = ParserState.DATA
          }
          break

        case ParserState.IAC_DO:
          this.emit('iacDo', byte)
          this.state = ParserState.DATA
          break

        case ParserState.IAC_DONT:
          this.emit('iacDont', byte)
          this.state = ParserState.DATA
          break

        case ParserState.IAC_WILL:
          this.emit('iacWill', byte)
          this.state = ParserState.DATA
          break

        case ParserState.IAC_WONT:
          this.emit('iacWont', byte)
          this.state = ParserState.DATA
          break

        case ParserState.IAC_SB:
          this.sbOption = byte
          this.sbData = []
          this.state = ParserState.IAC_SB_DATA
          break

        case ParserState.IAC_SB_DATA:
          if (byte === IAC) {
            this.state = ParserState.IAC_SB_IAC
          } else {
            this.sbData.push(byte)
          }
          break

        case ParserState.IAC_SB_IAC:
          if (byte === IAC) {
            this.sbData.push(IAC)
            this.state = ParserState.IAC_SB_DATA
          } else if (byte === SE) {
            this.emit('iacSb', { option: this.sbOption, data: Buffer.from(this.sbData) })
            this.state = ParserState.DATA
          } else {
            this.emit('iacSb', { option: this.sbOption, data: Buffer.from(this.sbData) })
            this.state = ParserState.DATA
          }
          break
      }
    }

    if (this.cleanBytes.length) {
      const cleanBuf = Buffer.from(this.cleanBytes)
      this.cleanBytes = []
      const clean = decodeBufferForTerminal(cleanBuf)
      this.emit('data', clean)
      return clean
    }

    return ''
  }

  on<K extends keyof ProtocolEvents>(event: K, callback: EventCallback<ProtocolEvents[K]>): void {
    if (!this.callbacks[event]) {
      this.callbacks[event] = []
    }
    this.callbacks[event]!.push(callback)
  }

  off<K extends keyof ProtocolEvents>(event: K, callback: EventCallback<ProtocolEvents[K]>): void {
    const list = this.callbacks[event]
    if (list) {
      const idx = list.indexOf(callback as any)
      if (idx >= 0) list.splice(idx, 1)
    }
  }

  reset(): void {
    this.state = ParserState.DATA
    this.sbOption = 0
    this.sbData = []
  }

  private emit<K extends keyof ProtocolEvents>(event: K, value: ProtocolEvents[K]): void {
    const list = this.callbacks[event]
    if (list) {
      for (const cb of list) {
        (cb as EventCallback<any>)(value)
      }
    }
  }
}

export { IAC, DO, DONT, WILL, WONT, SB, SE, OPT_ECHO, OPT_SUPPRESS_GO_AHEAD, OPT_NAWS }

File: src-tauri/src/connection/protocol_parser.rs
// Telnet IAC 协议解析器状态机
// 对应 src/main/connection/protocol-parser.ts

// Telnet IAC 常量
pub const IAC: u8 = 255;
pub const DO: u8 = 253;
pub const DONT: u8 = 254;
pub const WILL: u8 = 251;
pub const WONT: u8 = 252;
pub const SB: u8 = 250;
pub const SE: u8 = 240;

// Telnet 选项常量
pub const OPT_ECHO: u8 = 1;
pub const OPT_SUPPRESS_GO_AHEAD: u8 = 3;
pub const OPT_NAWS: u8 = 31;

// 解析器状态
const STATE_DATA: u8 = 0;
const STATE_IAC: u8 = 1;
const STATE_IAC_DO: u8 = 2;
const STATE_IAC_DONT: u8 = 3;
const STATE_IAC_WILL: u8 = 4;
const STATE_IAC_WONT: u8 = 5;
const STATE_IAC_SB: u8 = 6;
const STATE_IAC_SB_DATA: u8 = 7;
const STATE_IAC_SB_IAC: u8 = 8;

/// Telnet IAC 协议事件
#[derive(Debug, Clone)]
pub enum ProtocolEvent {
    /// 干净数据（已剥离 IAC）
    Data(String),
    IacDo(u8),
    IacDont(u8),
    IacWill(u8),
    IacWont(u8),
    /// 子协商：option, data
    IacSb { option: u8, data: Vec<u8> },
}

/// Telnet IAC 状态机解析器
pub struct ProtocolParser {
    state: u8,
    sb_option: u8,
    sb_data: Vec<u8>,
    clean_bytes: Vec<u8>,
}

impl Default for ProtocolParser {
    fn default() -> Self {
        Self::new()
    }
}

impl ProtocolParser {
    pub fn new() -> Self {
        Self {
            state: STATE_DATA,
            sb_option: 0,
            sb_data: Vec::new(),
            clean_bytes: Vec::new(),
        }
    }

    /// 输入原始字节流，返回解析出的事件列表
    pub fn feed(&mut self, raw: &[u8]) -> Vec<ProtocolEvent> {
        let mut events = Vec::new();

        for &byte in raw {
            match self.state {
                STATE_DATA => {
                    if byte == IAC {
                        self.state = STATE_IAC;
                    } else {
                        self.clean_bytes.push(byte);
                    }
                }
                STATE_IAC => {
                    if byte == IAC {
                        // IAC IAC → 透传一个 0xFF
                        self.clean_bytes.push(IAC);
                        self.state = STATE_DATA;
                    } else {
                        match byte {
                            DO => self.state = STATE_IAC_DO,
                            DONT => self.state = STATE_IAC_DONT,
                            WILL => self.state = STATE_IAC_WILL,
                            WONT => self.state = STATE_IAC_WONT,
                            SB => self.state = STATE_IAC_SB,
                            _ => self.state = STATE_DATA,
                        }
                    }
                }
                STATE_IAC_DO => {
                    events.push(ProtocolEvent::IacDo(byte));
                    self.state = STATE_DATA;
                }
                STATE_IAC_DONT => {
                    events.push(ProtocolEvent::IacDont(byte));
                    self.state = STATE_DATA;
                }
                STATE_IAC_WILL => {
                    events.push(ProtocolEvent::IacWill(byte));
                    self.state = STATE_DATA;
                }
                STATE_IAC_WONT => {
                    events.push(ProtocolEvent::IacWont(byte));
                    self.state = STATE_DATA;
                }
                STATE_IAC_SB => {
                    self.sb_option = byte;
                    self.sb_data.clear();
                    self.state = STATE_IAC_SB_DATA;
                }
                STATE_IAC_SB_DATA => {
                    if byte == IAC {
                        self.state = STATE_IAC_SB_IAC;
                    } else {
                        self.sb_data.push(byte);
                    }
                }
                STATE_IAC_SB_IAC => {
                    if byte == IAC {
                        // SB 数据中 IAC IAC → 透传 0xFF
                        self.sb_data.push(IAC);
                        self.state = STATE_IAC_SB_DATA;
                    } else if byte == SE {
                        // 子协商结束
                        let data = std::mem::take(&mut self.sb_data);
                        let option = self.sb_option;
                        events.push(ProtocolEvent::IacSb { option, data });
                        self.state = STATE_DATA;
                    } else {
                        // 异常终止，仍然派发已收集的数据
                        let data = std::mem::take(&mut self.sb_data);
                        let option = self.sb_option;
                        events.push(ProtocolEvent::IacSb { option, data });
                        self.state = STATE_DATA;
                    }
                }
                _ => self.state = STATE_DATA,
            }
        }

        // 派发干净数据
        if !self.clean_bytes.is_empty() {
            let buf = std::mem::take(&mut self.clean_bytes);
            let clean = super::c1_convert::decode_buffer_for_terminal(&buf);
            if !clean.is_empty() {
                events.push(ProtocolEvent::Data(clean));
            }
        }

        events
    }

    pub fn reset(&mut self) {
        self.state = STATE_DATA;
        self.sb_option = 0;
        self.sb_data.clear();
        self.clean_bytes.clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_data_passthrough() {
        let mut p = ProtocolParser::new();
        let events = p.feed(b"hello");
        assert_eq!(events.len(), 1);
        if let ProtocolEvent::Data(s) = &events[0] {
            assert_eq!(s, "hello");
        } else {
            panic!("expected Data event");
        }
    }

    #[test]
    fn test_iac_do() {
        let mut p = ProtocolParser::new();
        let events = p.feed(&[IAC, DO, OPT_ECHO]);
        assert!(events.iter().any(|e| matches!(e, ProtocolEvent::IacDo(OPT_ECHO))));
    }

    #[test]
    fn test_iac_iac_escape() {
        let mut p = ProtocolParser::new();
        let events = p.feed(&[IAC, IAC]);
        assert_eq!(events.len(), 1);
        if let ProtocolEvent::Data(s) = &events[0] {
            assert_eq!(s.as_bytes(), &[IAC]);
        }
    }

    #[test]
    fn test_subnegotiation() {
        let mut p = ProtocolParser::new();
        let events = p.feed(&[IAC, SB, OPT_NAWS, 0, 24, IAC, SE]);
        assert!(events.iter().any(|e| matches!(e, ProtocolEvent::IacSb { option: OPT_NAWS, .. })));
    }
}
